text/0000-pointer-match.md

```diff
@@ -0,0 +1,331 @@
+
+To create a pointer to a field of a struct, there is currently no way in Rust
+that avoids creating a temporary reference. Since reference semantics are
+stricter, this may lead to silent undefined behaviour where that reference
+should not be valid. Depending on the resolution of reference semantics this
+affects:
+
+* Creating a pointer to a field of a packed struct, where the reference may be
+  unaligned (depending on .
+* Pointing to fields of an uninitialized type, where the reference points to
+  uninitialized data. This may be complicated by unions, where it could be
+  possible that not a single variant is currently completely initialized, yet
+  one wants to access some subfield. See
+  <https://github.com/rust-lang/unsafe-code-guidelines/issues/73#issuecomment-460634637>.
+* Doing pointer offset calculations where the references does not refer to the
+  same, or any, allocation. This is because reference calculations are
+  performed with `getelementptr inbounds`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Match expression are extended from support for a reference binding mode, to a
+pointer binding mode. Furthermore, a new pattern binds to a pointer, and
+identifiers are extended to allow a new mode similar to `ref` and `ref mut`
+binding to a reference. These patterns are called pointer pattern and raw
+identifier for the remainder of the document.
+
+```
+#[repr(packed)]
+struct Foo {
+    a: u16,
+    b: u32,
+}
+
+fn ptr_b(foo: &mut Foo) -> *mut u32 {
+    let Foo { raw mut b, .. } = foo;
+    b
+}
+```
+
+Note that pointer binding mode and pointer pattern requires `unsafe`, even when
+it will never dereference the pointer. But the arithmetic on the pointer may
+implicitely overflow. Furthermore, not all patterns are (yet) allowed, to avoid
+implicitely performing an unintended, unsafe read through the pointer. Pointer
+binding mode will at first only permit ultimately binding with `raw` and `ref`
+and not actually reading the contained memory.
+
+The raw identifier pattern does not require `unsafe` on its own (as seen above,
+where we safely match a `&mut` but bind to `*mut`).
+
+This is not only useful for packed fields, but also to access the fields of
+*any* object that is only available via pointer because its state invariants
+may not yet be fulfilled. Instead of manually doing pointer math:
+
+```
+/// Repr Rust, so no layout guarantees, no pointer operations to get to `a`.
+struct Weird {
+    /// Always valid, no matter the memory content.
+    something_i_dont_care_about: u8,
+
+    /// Must only be one of `true` and `false` for a &Weird.
+    a: bool,
+}
+
+/// Unsafety invariants: `w` must 
+/// * point to some allocation of at least `std::mem::size_of::<Weird>`.
+/// * point to memory valid for the chosen lifetime `'a`
+/// * be properly aligned.
+unsafe fn get_if_init<'a>(w: *const Weird) -> Option<&'a Weird> {
+    let Weird { raw const a, ..} = w;
```

Comment:
Does this auto-deref, or how does this typecheck?

Your example below with `match (0 as *mut usize)` does not look like auto-deref happens, but then shouldn't this be `let *const Weird { raw const a, .. } = w`?

Reply:
`match` doesn't auto-deref in the `ops::Deref` sense of expressions. I'd rather view it as only adding reference patterns automatically where required, [see the implementation](https://github.com/rust-lang/rust/blob/3752b3d3a56bf3eabb588b7d595cd1f8accc0286/src/librustc_mir/hair/pattern/mod.rs#L365) (otherwise [somethink like this could be argued to work](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5282a7f0af2d2b9a6efa98506c6829d9) and be a horror to match). I only suggest automatically adding pointer patterns if required. Which pattern to add should never be ambiguous–in a top-down pass of the pattern after the rhs of the match is typed–but probably deserves its own section anyways.

Reply:
> Your example below with match (0 as *mut usize) does not look like auto-deref happens

Automatically adding reference for pointers has the slightly unfortunate side-effect of colliding with the ability to match pointers by const value (i.e. `core::ptr::null()`), which does not exist for references afaik. The example was intended to show that disambiguation should be backwards compatible but didn't show automatically adding the pattern.

Reply:
> match doesn't auto-deref in the ops::Deref sense of expressions. I'd rather view it as only adding reference patterns automatically where required

I do not see how that makes any fundamental difference. The fact remains taht you can write a pointer deref and incur a memory access without ever typing `*`, such as in
```rust
fn foo(x: &bool) -> bool {
    match x {
        true => true,
        false => false,
    }
}
```
For raw pointers, we want to avoid this because accessing memory through them is unsafe and should only be done explicitly. In this sense, auto-deref on raw pointers and auto-`&`-pattern in match have the same concern.

This is not about ambiguity, this is about calling out to whoeever reads this code that *a raw pointer is being dereferenced*.  I don't think this should compile:
```rust
fn foo(x: *const bool) -> bool {
    unsafe { match x {
        true => true,
        false => false,
    } }
}
```

Reply:
That code would not work. While the pointer pattern is *added* automatically, it still does not allow any of the reading patterns to occur within it. So the value pattern `true` within in that example would desugar to `*const true` and then be denied on the grounds that a value pattern occurred inside a pointer pattern. I believe that the wording correctly captures this part of the mechanic but I could have misphrased it. Hence the distinction of `auto-deref` and adding a pattern, because the latter does not circumvent the pattern-soundness checks.

Reply:
That would indeed fix this concern. It wasn't clear to me from the RFC; there should be examples for this kind of rule.